streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -810,10 +810,19 @@ private void transitRestoredTaskToRunning(final Task task,
     }
 
     private void addTasksToStateUpdater() {
+        final Map<TaskId, RuntimeException> taskExceptions = new LinkedHashMap<>();
         for (final Task task : tasks.drainPendingTaskToInit()) {
-            task.initializeIfNeeded();
-            stateUpdater.add(task);
+            try {
+                task.initializeIfNeeded();
+                stateUpdater.add(task);
+            } catch (final RuntimeException e) {
+                // need to add task back to the bookkeeping to be handled by the stream thread
+                tasks.addTask(task);
```

Comment:
I am wondering if it wouldn't be better to add the entire task to the `TaskCorruptedException` instead of only the task ID. In such a way, we wouldn't need to add a corrupted task back to the `TasksRegistry`. Is there somewhere a check that avoids executing corrupted tasks? Before the state updater we did not execute tasks before all tasks were restored but that is not the case anymore with the state updater.
Modifying the `TaskCorruptedException` would need a KIP. Instead of creating a KIP for modifying the `TaskCorruptedException`, we should create one to move `TaskCorruptedException` and `TaskMigratedException` to internal packages since they are completely handled within Streams and they should never be handled by the user.

Reply:
FWIW w.r.t the idea of attaching the entire task instead of just its taskId, I agree with Lucas here -- seems risky, would likely end up being abused as things progress (by us not the users), but most importantly imo is that it's just bad practice. I'm not too well versed on anti patterns I'll admit (perks & downsides of being a physics major 😉), but I would think this counts as one (or should) -- we shouldn't use exceptions to pass around actual objects rather than simple info/metadata. Especially objects of uncertain status, questionable thread safety, and significant resources attached. 

Just my two cents as someone who's coming into the restoration thread game very late -- given this fact I'm not particularly qualified to weigh in on what the _right_ thing to do is here, though again I feel like Lucas's suggestion to maintain a/the `TaskRegistry` as a central, source-of-truth type thing for tracking and organizing all the tasks

Reply:
I think I agree with the risk of passing the entire task to an exception that does not have an explicit way how to cleanup itself like the `shutdown()`methods in `TaskManager` and `StateUpdater`. However, I am not a fan of a central place for managing all tasks. We explicitly decided to have the invariant that a task that is managed by the state updater should not exist in the task manager and vice versa. That is to avoid concurrent modifications of the tasks by both threads. 
We need to evaluate if in the places where all tasks are returned really all tasks need to be returned and if yes we also need to call `getTasks()` on the state updater.

Reply:
Regarding the guarantee of not processing corrupted tasks, I think we should be safe since `maybeThrowTaskExceptions()` immediately throws the encountered exceptions. I just dislike the fact that we rely on the exceptions bubbling up all the way up to `StreamThread#runLoop()` which might be broken by future changes.